### ISSUE TITLE
Arithmetic overflow in particle ID

### ISSUE THREAD

**Issue by lucafedeli88** [ECP-WarpX/WarpX#643]
Hi,
I've just found a bug related to an overflow of the particle ID. 

I was experimenting with photon emission in my QED branch and I was doing a simulation with just one MPI task and one openMP thread. I was using a 1 um x 1 um box with only a tiny region (~ 0.02 um x 0.02 um) filled with 10.000 electrons per cell. I noticed that changing the resolution from 256x256 to 512x512 the number of generated photons plummeted to exactly zero (while at 256x256 I got very reasonable results for my physical case). 

I discovered that code responsible for photon generation was actually called, but photons were assigned negative IDs (and thus immediately destroyed).

I've just found out that the following lines in `PhysicalParticleContainer.cpp` are responsible for an arithmetic overflow, which leads to the negative IDs :
```
pid = ParticleType::NextID();
ParticleType::NextID(pid+max_new_particles);
```
I suspect that in this case `max_new_particles` greatly overestimates the number of particles to be generated, due to the very high number of particles per cell that I am using.

We might consider to add a check to ensure that `pid+max_new_particles > 0` or something similar...

**Comment by atmyers** [comment]
I believe the fix here is to switch to a two-pass approach in AddPlasma, where we first count the number of particles to add, and then add exactly that number, instead of creating invalid particles and then relying on the Redistribute to remove them. This will also be useful for @RevathiJambunathan's pulsar project.

I'll make a PR to do this.

**Comment by WeiqunZhang** [comment]
Are there cases the two-pass approach will not work because the number of particles to be added is random and the second pass won't add the same number particles?

**Comment by atmyers** [comment]
We can handle that by building a mask or something and reusing it between the two passes. We could only do that if it's needed.

**Comment by atmyers** [comment]
Actually, maybe there is a simpler way to do this. Currently, there is an early exit in the MFIter loop that generates particles if there is no way that tile could intersect with the plasma region. Only if there is possible overlap do we advance `NextID`, allocate space, etc.

What if we make this loop execute with a tile size that depends on num_ppc (even on GPUs), so as to exclude more of the volume that is outside the plasma region?

**Comment by atmyers** [comment]
Alternatively, a two-pass approach that could handle random is the following. First, count the number of cells that overlap with the region. Then, allocate enough space / advance NextID for num_cells_inside*num_pcc particles. Then, launch the kernel that actually generates positions and fills in the data. Particles that end up outside at this point could be invalidated as usually, and the amount of excess would be quite small.

**Comment by atmyers** [comment]
@lucafedeli88 Do you happen to have an inputs script that reproduces this issue? I have a potential fix and I'd like to test it.

**Comment by RemiLehe** [comment]
This was probably fixed by #830. Is that correct @lucafedeli88 ?

**Comment by lucafedeli88** [comment]
Ah, yes! We can close this issue.
